### PR TITLE
Store/containerized apps support

### PR DESCRIPTION
**Pull request by tiagoshibata**
**Description**: Support buliding ORT and WinML for Windows Store and containerized apps.

This PR removes references to functions unavailable in appcontainer/Store. I'll do another PR with CMake changes to target Store apps.

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -162,11 +162,17 @@ if(NOT WIN32)
...
-    # cmake/external/protobuf/src/google/protobuf/compiler/subprocess.cc and onnxruntime/core/platform/windows/env.cc call a bunch of Win32 APIs.
-    # For now, we'll set the API family to desktop globally to expose Win32 symbols in headers; this must be fixed!
-    add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:/FI${CMAKE_CURRENT_SOURCE_DIR}/set_winapi_family_desktop.h>")
+    if (WINDOWS_STORE_FOR_REAL)
+#      add_compile_definitions(WINAPI_FAMILY=2 __cplusplus_winrt)
+#      add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:SHELL:/FIvccorlib.h /ZW:nostdlib /FUplatform.winmd>")
+      add_compile_definitions(WINAPI_FAMILY=2)
+      add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:SHELL:/ZW:nostdlib /FUplatform.winmd>")
+
+      add_link_options(/APPCONTAINER)
```
Comment (tiagoshibata) [inline cmake/CMakeLists.txt:172]: This is not duplicated; both `CMAKE_VS_GLOBALS "AppContainerApplication=true"` and `add_link_options(/APPCONTAINER)` are needed. The latter is needed for Ninja builds (it won't use the VS project properties)

File: onnxruntime/core/platform/windows/env.cc [hunk 2]
```diff
@@ -397,6 +380,10 @@ class WindowsEnv : public Env {
   common::Status GetCanonicalPath(
       const PathString& path,
       PathString& canonical_path) const override {
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
+    canonical_path = path;
+    return Status::OK();
+#else
```
Comment (tiagoshibata) [inline onnxruntime/core/platform/windows/env.cc:386]: I'll revisit this later to see if there's any APIs to do this in an appcontainer

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 3]
```diff
@@ -510,6 +510,9 @@ struct ProviderLibrary {
       return;
 
 #if defined(_WIN32) && !defined(_OPENMP)
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
+#error Unsupported configuration
```
Comment (martinb35) [inline onnxruntime/core/framework/provider_bridge_ort.cc:514]: can you put a comment on why this is unsupported?

Comment (tiagoshibata) [reply onnxruntime/core/framework/provider_bridge_ort.cc:514]: I'm not too sure why this DNNL/OMP workaround is needed in the first place, I'm asking ORT here: https://github.com/microsoft/onnxruntime/pull/3978#issuecomment-667342817 . Maybe there's a way to make it work without using GetModuleHandleExA

File: onnxruntime/core/platform/windows/env.cc [hunk 4]
```diff
@@ -67,7 +67,9 @@ class WindowsThread : public EnvThread {
   // This function is called when the threadpool is cancelled.
   // TODO: Find a way to avoid calling TerminateThread
   void OnCancel() {
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env.cc:70]: In the store context the thread will keep running?

This seems like it could potentially cause problems?

File: onnxruntime/core/platform/windows/env.cc [hunk 5]
```diff
@@ -471,7 +494,11 @@ class WindowsEnv : public Env {
   }
 
   virtual Status LoadDynamicLibrary(const std::string& library_filename, void** handle) const override {
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
+    *handle = ::LoadPackagedLibrary(ToWideString(library_filename).c_str(), 0);
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env.cc:498]: When calling the UWP build dll in a desktop application, now we have different behavior where dlls can potentially be loaded from the application directory rather than the module directory. (with a different search order)

For UWP apps this is probably okay as the search order should be controlled by the sandbox.

File: winml/lib/Api.Ort/OnnxruntimeEnvironment.cpp [hunk 6]
```diff
@@ -14,12 +14,16 @@ using namespace _winml;
 static bool debug_output_ = false;
 
 static HRESULT GetOnnxruntimeLibrary(HMODULE& module) {
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
+  auto out_module = LoadPackagedLibrary(L"onnxruntime.dll", 0);
```
Comment (smk2007) [inline winml/lib/Api.Ort/OnnxruntimeEnvironment.cpp:18]: Issues when called in a desktop context:
Desktop applications cannot use LoadPackagedLibrary; if a desktop application calls this function it fails with APPMODEL_ERROR_NO_PACKAGE.

File: onnxruntime/core/platform/windows/env.cc (hunk 5 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:498]: But we'll have a separate build for Store app and desktop. The Store build won't run as an uncontainerized desktop app.

File: onnxruntime/core/platform/windows/env.cc (hunk 4 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:70]: Yes, there's no way to TerminateThread in UWP and it will keep running to finish its job. Though this shouldn't be called anyways (you should finish all threads before cancelling the threadpool, and terminating the thread leaks handles/memory/mutexes/whatever the thread was holding).

This will also match ORT's behavior elsewhere (https://github.com/microsoft/onnxruntime/blob/01ca6392cb165f3ebad6fc3c28016adad5a26688/onnxruntime/core/platform/posix/env.cc#L157-L160 does nothing when you cancel a running thread).

File: onnxruntime/core/platform/windows/env.cc (hunk 5 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:498]: There's an `#else` branch for the non-app case, that calls LoadLibraryExA

Comment (smk2007) [reply onnxruntime/core/platform/windows/env.cc:498]: WinUI only creates a Store build, and then relies on the vcforwarders to have the store binaries run everywhere.

If possible we should ensure that we follow a similar principle.

Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:498]: There isn't a common set of APIs that goes from Desktop to UWP to load libraries during runtime. I think that would only be possible if we delayload LoadPackagedLibrary/LoadLibraryExA depending on the platform (and CreateFile2/CreateFileW, GetModuleHandle, SetThreadDescription...).

I didn't know about vcforwarders, are you talking about this project: https://github.com/microsoft/vcrt-forwarders ? Do we want to support the use case of running a Store build on desktop (or vice-versa) and just work?

File: winml/lib/Api.Ort/OnnxruntimeErrors.h [hunk 7]
```diff
@@ -38,17 +38,24 @@ inline HRESULT OrtErrorCodeToHRESULT(OrtErrorCode status) noexcept {
...
     if (_status) {                                                                                             \
       auto error_code = ort_api->GetErrorCode(_status);                                                        \
       auto error_message = ort_api->GetErrorMessage(_status);                                                  \
       HRESULT hresult = OrtErrorCodeToHRESULT(error_code);                                                     \
-      telemetry_helper.LogRuntimeError(hresult, std::string(error_message), __FILE__, __FUNCTION__, __LINE__); \
-      auto message = _winml::Strings::HStringFromUTF8(error_message);                                           \
-      RoOriginateError(hresult, reinterpret_cast<HSTRING>(winrt::get_abi(message)));                           \
-      return hresult;                                                                                          \
+      telemetry_helper.LogRuntimeError(hresult, error_message, __FILE__, __FUNCTION__, __LINE__); \
+      return OriginateError(hresult, error_message);                                                           \
```
Comment (tiagoshibata) [inline winml/lib/Api.Ort/OnnxruntimeErrors.h:58]: I'm piggybacking some downlevel fixes that showed up with the new WINVER macros. RoOriginateError is >= Win8

File: onnxruntime/core/platform/windows/env.cc [hunk 8]
```diff
@@ -179,8 +185,13 @@ class WindowsEnv : public Env {
   }
 
   Status GetFileLength(_In_z_ const ORTCHAR_T* file_path, size_t& length) const override {
+#if WINVER >= _WIN32_WINNT_WIN8
     wil::unique_hfile file_handle{
-        CreateFileW(file_path, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL)};
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env.cc:183]: >GENERIC_READ [](start = 31, length = 12)

why switch to FILE_READ_ATTRIBUTES?

File: onnxruntime/core/platform/windows/env.cc [hunk 9]
```diff
@@ -399,7 +414,14 @@ class WindowsEnv : public Env {
       PathString& canonical_path) const override {
     // adapted from MSVC STL std::filesystem::canonical() implementation
     // https://github.com/microsoft/STL/blob/ed3cbf36416a385828e7a5987ca52cb42882d84b/stl/inc/filesystem#L2986
-
+#if WINVER >= _WIN32_WINNT_WIN8
+    wil::unique_hfile file_handle{CreateFile2(
+        path.c_str(),
+        FILE_READ_ATTRIBUTES,
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env.cc:420]: >FILE_READ_ATTRIBUTES [](start = 8, length = 20)

seems like an inconsistent use of FILE_READ_ATTRIBUTES and GENERIC_READ

File: onnxruntime/core/platform/windows/env.cc (hunk 8 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:183]: It's only reading the length instead of contents

File: onnxruntime/core/platform/windows/env_time.cc [hunk 10]
```diff
@@ -28,9 +28,12 @@ namespace {
 
 class WindowsEnvTime : public EnvTime {
  public:
+#if WINVER >= _WIN32_WINNT_WIN8
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env_time.cc:31]: This isnt a runtime check? Was it intended to be?

File: onnxruntime/core/platform/windows/env.cc (hunk 8 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env.cc:183]: And that's how it was done elsewhere (in the `GetCanonicalPath` function)

File: onnxruntime/core/platform/windows/env_time.cc [hunk 11]
```diff
@@ -28,9 +28,12 @@ namespace {
...
+  WindowsEnvTime() : GetSystemTimePreciseAsFileTime_(GetSystemTimePreciseAsFileTime) {}
+#else
   WindowsEnvTime() : GetSystemTimePreciseAsFileTime_(NULL) {
-    // GetSystemTimePreciseAsFileTime function is only available in the latest
-    // versions of Windows. For that reason, we try to look it up in
+    // GetSystemTimePreciseAsFileTime function is only available in
+    // Windows >= 8. For that reason, we try to look it up in
```
Comment (smk2007) [inline onnxruntime/core/platform/windows/env_time.cc:36]: >Windows >= 8. [](start = 7, length = 13)

Do we need win7 support here? My guess is this is related to wanting telemetry for the edge builds?

File: onnxruntime/core/platform/windows/env_time.cc (hunk 10 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env_time.cc:31]: No, it's intended to be a compile time check. If it's a build that must work downlevel (normal desktop build) it will call `GetModuleHandleW` to find `GetSystemTimePreciseAsFileTime` during runtime. If it's supposed to run in an appcontainer (which is always >= Win10, and always has `GetSystemTimePreciseAsFileTime`), it builds with GetSystemTimePreciseAsFileTime.

File: onnxruntime/core/platform/windows/env_time.cc (hunk 11 above)
Comment (tiagoshibata) [reply onnxruntime/core/platform/windows/env_time.cc:36]: I thought we wanted Win7 support in ORT/WinML?

File: winml/lib/Api.Ort/OnnxruntimeErrors.h [hunk 12]
```diff
@@ -38,17 +38,24 @@ inline HRESULT OrtErrorCodeToHRESULT(OrtErrorCode status) noexcept {
 }
 #endif
 
+static inline HRESULT OriginateError(HRESULT hresult, const char *message) {
+#if WINVER >= _WIN32_WINNT_WIN8
```
Comment (smk2007) [inline winml/lib/Api.Ort/OnnxruntimeErrors.h:42]: winml isnt supported on win7 - why was this needed?

**Review by smk2007 — DISMISSED** [review]

File: onnxruntime/core/platform/windows/env.cc [hunk 13]
```diff
@@ -78,8 +80,12 @@ class WindowsThread : public EnvThread {
     if (!p->thread_options.affinity.empty())
       SetThreadAffinityMask(GetCurrentThread(), p->thread_options.affinity[p->index]);
     // kernel32.dll is always loaded
+#if WINVER >= _WIN32_WINNT_WIN10
+    constexpr SetThreadDescriptionFunc pSetThrDesc = SetThreadDescription;
+#else
```
Comment (tiagoshibata) [inline onnxruntime/core/platform/windows/env.cc:85]: FIXME: Store Windows 8 support

**Review by martinb35 — APPROVED** [review]

**Review by martinb35 — DISMISSED** [review]

File: winml/test/common/googleTestMacros.h [hunk 14]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
-
+#define _SILENCE_CXX17_CODECVT_HEADER_DEPRECATION_WARNING
```
Comment (ryanlai2) [inline winml/test/common/googleTestMacros.h:3]: Which part of test code needs to silence deprecation warnings? Does this also need to be in taefTestMacros.h?

Comment (tiagoshibata) [reply winml/test/common/googleTestMacros.h:3]: Line 87. Not really needed for now since warnings as errors aren't enabled, but I'll make another PR in the future fixing remaining warnings.

**Review by martinb35 — APPROVED** [review]
